File: app/src/main/java/com/kickstarter/ui/viewholders/ProfileCardViewHolder.java
package com.kickstarter.ui.viewholders;

import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.kickstarter.R;
import com.kickstarter.libs.utils.ObjectUtils;
import com.kickstarter.libs.utils.ProgressBarUtils;
import com.kickstarter.models.Photo;
import com.kickstarter.models.Project;
import com.squareup.picasso.Picasso;

import butterknife.Bind;
import butterknife.BindDrawable;
import butterknife.BindString;
import butterknife.ButterKnife;

public final class ProfileCardViewHolder extends KSViewHolder {
  private final Delegate delegate;
  private Project project;

  protected @Bind(R.id.funding_unsuccessful_text_view) TextView fundingUnsuccessfulTextView;
  protected @Bind(R.id.percentage_funded) ProgressBar percentageFundedProgressBar;
  protected @Bind(R.id.profile_card_image) ImageView profileCardImageView;
  protected @Bind(R.id.profile_card_name) TextView profileCardNameTextView;
  protected @Bind(R.id.project_state_view_group) ViewGroup projectStateViewGroup;
  protected @Bind(R.id.successfully_funded_text_view) TextView successfullyFundedTextView;

  protected @BindDrawable(R.drawable.gray_gradient) Drawable grayGradientDrawable;

  protected @BindString(R.string.profile_projects_status_successful) String successfulString;
  protected @BindString(R.string.profile_projects_status_unsuccessful) String unsuccessfulString;
  protected @BindString(R.string.profile_projects_status_canceled) String cancelledString;
  protected @BindString(R.string.profile_projects_status_suspended) String suspendedString;

  public interface Delegate {
    void profileCardViewHolderClicked(ProfileCardViewHolder viewHolder, Project project);
  }

  public ProfileCardViewHolder(final @NonNull View view, final @NonNull Delegate delegate) {
    super(view);
    this.delegate = delegate;
    ButterKnife.bind(this, view);
  }

  @Override
  public void bindData(final @Nullable Object data) throws Exception {
    this.project = ObjectUtils.requireNonNull((Project) data, Project.class);
  }

  @Override
  public void onBind() {
    final Photo photo = this.project.photo();

    if (photo != null) {
      this.profileCardImageView.setVisibility(View.VISIBLE);
      Picasso.get().load(photo.med())
        .placeholder(this.grayGradientDrawable)
        .into(this.profileCardImageView);
    } else {
      this.profileCardImageView.setVisibility(View.INVISIBLE);
    }

    this.profileCardNameTextView.setText(this.project.name());
    this.percentageFundedProgressBar.setProgress(ProgressBarUtils.progress(this.project.percentageFunded()));

    setProjectStateView();
  }

  @Override
  public void onClick(final @NonNull View view) {
    this.delegate.profileCardViewHolderClicked(this, this.project);
  }

  public void setProjectStateView() {
    switch(this.project.state()) {
      case Project.STATE_SUCCESSFUL:
        this.percentageFundedProgressBar.setVisibility(View.GONE);
        this.projectStateViewGroup.setVisibility(View.VISIBLE);
        this.fundingUnsuccessfulTextView.setVisibility(View.GONE);
        this.successfullyFundedTextView.setVisibility(View.VISIBLE);
        this.successfullyFundedTextView.setText(this.successfulString);
        break;
      case Project.STATE_CANCELED:
        this.percentageFundedProgressBar.setVisibility(View.GONE);
        this.projectStateViewGroup.setVisibility(View.VISIBLE);
        this.successfullyFundedTextView.setVisibility(View.GONE);
        this.fundingUnsuccessfulTextView.setVisibility(View.VISIBLE);
        this.fundingUnsuccessfulTextView.setText(this.cancelledString);
        break;
      case Project.STATE_FAILED:
        this.percentageFundedProgressBar.setVisibility(View.GONE);
        this.projectStateViewGroup.setVisibility(View.VISIBLE);
        this.successfullyFundedTextView.setVisibility(View.GONE);
        this.fundingUnsuccessfulTextView.setVisibility(View.VISIBLE);
        this.fundingUnsuccessfulTextView.setText(this.unsuccessfulString);
        break;
      case Project.STATE_SUSPENDED:
        this.percentageFundedProgressBar.setVisibility(View.GONE);
        this.projectStateViewGroup.setVisibility(View.VISIBLE);
        this.successfullyFundedTextView.setVisibility(View.GONE);
        this.fundingUnsuccessfulTextView.setVisibility(View.VISIBLE);
        this.fundingUnsuccessfulTextView.setText(this.suspendedString);
        break;
      default:
        this.percentageFundedProgressBar.setVisibility(View.VISIBLE);
        this.projectStateViewGroup.setVisibility(View.GONE);
        break;
    }
  }
}

File: app/src/main/java/com/kickstarter/ui/viewholders/ProfileCardViewHolder.kt
package com.kickstarter.ui.viewholders

import android.view.View
import com.kickstarter.R
import com.kickstarter.databinding.ProfileCardViewBinding
import com.kickstarter.libs.utils.ObjectUtils
import com.kickstarter.libs.utils.ProgressBarUtils
import com.kickstarter.models.Project
import com.squareup.picasso.Picasso

class ProfileCardViewHolder(
    private val binding: ProfileCardViewBinding,
    private val delegate: Delegate
) : KSViewHolder(binding.root) {
    private var project: Project? = null

    interface Delegate {
        fun profileCardViewHolderClicked(viewHolder: ProfileCardViewHolder, project: Project)
    }

    @Throws(Exception::class)
    override fun bindData(data: Any?) {
        project = ObjectUtils.requireNonNull(data as Project?, Project::class.java)
    }

    override fun onBind() {
        val photo = project?.photo()

        binding.profileCardImage.visibility = View.INVISIBLE

        if (photo != null) {
            binding.profileCardImage.apply {
                visibility = View.VISIBLE
                Picasso.with(context()).load(photo.med())
                    .placeholder(context().getDrawable(R.drawable.gray_gradient))
                    .into(this)
            }
        }
        binding.profileCardName.text = project?.name()
        project?.percentageFunded()?.let {
            binding.percentageFunded.progress = ProgressBarUtils.progress(it)
        }
        setProjectStateView()
    }

    override fun onClick(view: View) {
        project?.let { delegate.profileCardViewHolderClicked(this, it) }
    }

    fun setProjectStateView() {
        when (project?.state()) {
            Project.STATE_SUCCESSFUL -> {
                binding.percentageFunded.visibility = View.GONE
                binding.projectStateViewGroup.visibility = View.VISIBLE
                binding.fundingUnsuccessfulTextView.visibility = View.GONE
                binding.successfullyFundedTextView.visibility = View.VISIBLE
                binding.successfullyFundedTextView.text = context().getString(R.string.profile_projects_status_successful)
            }
            Project.STATE_CANCELED -> {
                binding.percentageFunded.visibility = View.GONE
                binding.projectStateViewGroup.visibility = View.VISIBLE
                binding.successfullyFundedTextView.visibility = View.GONE
                binding.fundingUnsuccessfulTextView.visibility = View.VISIBLE
                binding.fundingUnsuccessfulTextView.text = context().getString(R.string.profile_projects_status_canceled)
            }
            Project.STATE_FAILED -> {
                binding.percentageFunded.visibility = View.GONE
                binding.projectStateViewGroup.visibility = View.VISIBLE
                binding.successfullyFundedTextView.visibility = View.GONE
                binding.fundingUnsuccessfulTextView.visibility = View.VISIBLE
                binding.fundingUnsuccessfulTextView.text = context().getString(R.string.profile_projects_status_unsuccessful)
            }
            Project.STATE_SUSPENDED -> {
                binding.percentageFunded.visibility = View.GONE
                binding.projectStateViewGroup.visibility = View.VISIBLE
                binding.successfullyFundedTextView.visibility = View.GONE
                binding.fundingUnsuccessfulTextView.visibility = View.VISIBLE
                binding.fundingUnsuccessfulTextView.text = context().getString(R.string.profile_projects_status_suspended)
            }
            else -> {
                binding.percentageFunded.visibility = View.VISIBLE
                binding.projectStateViewGroup.visibility = View.GONE
            }
        }
    }
}
